http-clients/netty-nio-client/src/main/java/software/amazon/awssdk/http/nio/netty/internal/http2/Http2ToHttpInboundAdapter.java

```diff
@@ -54,7 +59,21 @@ protected void channelRead0(ChannelHandlerContext ctx, Http2Frame frame) throws
     }
 
     private void onHeadersRead(Http2HeadersFrame headersFrame, ChannelHandlerContext ctx) throws Http2Exception {
-        ctx.fireChannelRead(HttpConversionUtil.toHttpResponse(headersFrame.stream().id(), headersFrame.headers(), true));
+
+        HttpResponse httpResponse = HttpConversionUtil.toHttpResponse(headersFrame.stream().id(), headersFrame.headers(), true);
+        ctx.fireChannelRead(httpResponse);
+
+        if (HttpStatusFamily.of(httpResponse.status().code()) == HttpStatusFamily.SERVER_ERROR) {
+            fireConnectionExceptionForServerError(ctx);
+        }
+    }
+
+    private void fireConnectionExceptionForServerError(ChannelHandlerContext ctx) {
+        if (ctx.channel().parent() != null) {
+            Channel parent = ctx.channel().parent();
+            log.debug(() -> "A 5xx server error occurred on an Http2 stream, notifying the connection channel " + ctx.channel());
+            parent.pipeline().fireExceptionCaught(Http2StreamServerErrorException.INSTANCE);
```

Comment:
Do we not support modeled 5xx exceptions for HTTP2 connections? It feels like this would break them, if we did.

Reply:
Actually, looks like this may not close the connection immediately, just closes it to new requests. I'm not sure if I like the instanceof checks to disambiguate this IOException from other IOExceptions. Should we have a different root exception type?

Reply:
We support modeled 5xx exceptions in core, and http client modules do not have knowledge of it.

Yeah, it doesn't close the connection immediately; the `RecordState` will be changed to `CLOSE_TO_NEW`

I can change it to `RuntimeException`. WDUT?

Reply:
So, all RuntimeExceptions on the channel will put it into no-new-connections state? I don't think that's a bad thing, but wanted to make sure my proposal was clear.

Reply:
So,  `Http2StreamServerErrorException` (extends RuntimeException) or `Http2StreamIoException`(extends IOException) on the stream channel will put the connection channel into close_to_new state. 

All other exceptions on the stream channels will not have any effect on connection channel.

Reply:
Discussed offline. We decided to have a single exception class for both server errors and io error on streams so that we don't need to update the exception handling logic if there are more cases in the future